File: app/src/main/java/com/checkin/app/checkin/Menu/MenuRepository.java
package com.checkin.app.checkin.Menu;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;

import com.checkin.app.checkin.Data.ApiClient;
import com.checkin.app.checkin.Data.ApiResponse;
import com.checkin.app.checkin.Data.AppDatabase;
import com.checkin.app.checkin.Data.NetworkBoundResource;
import com.checkin.app.checkin.Data.ObjectBoxInstanceLiveData;
import com.checkin.app.checkin.Data.Resource;
import com.checkin.app.checkin.Data.RetrofitLiveData;
import com.checkin.app.checkin.Data.WebApiService;
import com.checkin.app.checkin.Menu.Model.MenuModel;
import com.checkin.app.checkin.Menu.Model.MenuModel_;
import com.checkin.app.checkin.Menu.Model.OrderedItemModel;
import com.fasterxml.jackson.databind.node.ArrayNode;

import java.util.List;

import javax.inject.Singleton;

import io.objectbox.Box;

@Singleton
public class MenuRepository {
    private static MenuRepository INSTANCE;
    private Box<MenuModel> mMenuBox;
    private WebApiService mWebService;

    private MenuRepository(Context context) {
        mWebService = ApiClient.getApiService(context);
        mMenuBox = AppDatabase.getMenuModel(context);
    }

    public static MenuRepository getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (MenuRepository.class) {
                if (INSTANCE == null) {
                    Context context = application.getApplicationContext();
                    INSTANCE = new MenuRepository(context);
                }
            }
        }
        return INSTANCE;
    }

    public LiveData<Resource<MenuModel>> getAvailableMenu(final long shopId) {
        return new NetworkBoundResource<MenuModel, MenuModel>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @Override
            protected boolean shouldFetch(MenuModel data) {
                return true;
            }

            @Override
            protected LiveData<MenuModel> loadFromDb() {
                return new ObjectBoxInstanceLiveData<>(mMenuBox
                        .query().equal(MenuModel_.restaurantPk, shopId)
                        .build());
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<MenuModel>> createCall() {
                return new RetrofitLiveData<>(mWebService.getAvailableMenu(shopId));
            }

            @Override
            protected void saveCallResult(MenuModel data) {
                if (data != null) {
                    data.setRestaurantPk(shopId);
                    mMenuBox.put(data);
                }
            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ArrayNode>> postMenuOrders(final List<OrderedItemModel> orders) {
        return new NetworkBoundResource<ArrayNode, ArrayNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ArrayNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.postActiveSessionOrders(orders));
            }

            @Override
            protected void saveCallResult(ArrayNode data) {

            }
        }.getAsLiveData();
    }

    public LiveData<Resource<ArrayNode>> postMenuManageOrders(final long sessionPk, final List<OrderedItemModel> orders) {
        return new NetworkBoundResource<ArrayNode, ArrayNode>() {
            @Override
            protected boolean shouldUseLocalDb() {
                return false;
            }

            @NonNull
            @Override
            protected LiveData<ApiResponse<ArrayNode>> createCall() {
                return new RetrofitLiveData<>(mWebService.postSessionManagerOrders(sessionPk, orders));
            }

            @Override
            protected void saveCallResult(ArrayNode data) {

            }
        }.getAsLiveData();
    }
}

File: app/src/main/java/com/checkin/app/checkin/menu/MenuRepository.kt
package com.checkin.app.checkin.menu

import android.app.Application
import android.content.Context
import androidx.lifecycle.LiveData
import com.checkin.app.checkin.data.Converters
import com.checkin.app.checkin.data.db.AppDatabase
import com.checkin.app.checkin.data.db.ObjectBoxInstanceLiveData
import com.checkin.app.checkin.data.network.ApiClient.Companion.getApiService
import com.checkin.app.checkin.data.network.ApiResponse
import com.checkin.app.checkin.data.network.RetrofitLiveData
import com.checkin.app.checkin.data.network.WebApiService
import com.checkin.app.checkin.data.resource.NetworkBoundResource
import com.checkin.app.checkin.data.resource.Resource
import com.checkin.app.checkin.menu.models.*
import com.checkin.app.checkin.session.models.TrendingDishModel
import com.checkin.app.checkin.utility.SingletonHolder
import com.fasterxml.jackson.databind.node.ObjectNode
import io.objectbox.Box

class MenuRepository private constructor(context: Context) {
    private val mMenuBox: Box<MenuModel> = AppDatabase.getMenuModel(context)
    private val mCartBox: Box<CartStatusModel> = AppDatabase.getCartStatusModel(context)
    private val mWebService: WebApiService = getApiService(context)

    fun getAvailableMenu(shopId: Long): LiveData<Resource<MenuModel>> {
        return object : NetworkBoundResource<MenuModel, MenuModel>() {
            override fun shouldUseLocalDb(): Boolean = true

            override fun loadFromDb(): LiveData<MenuModel> {
                return ObjectBoxInstanceLiveData(mMenuBox
                        .query().equal(MenuModel_.restaurantPk, shopId)
                        .build())
            }

            override fun createCall(): LiveData<ApiResponse<MenuModel>> {
                return RetrofitLiveData(mWebService.getAvailableMenu(shopId))
            }

            override fun saveCallResult(data: MenuModel?) {
                data?.let {
                    it.restaurantPk = shopId
                    mMenuBox.put(it)
                }
            }
        }.asLiveData
    }

    fun postActiveSessionOrders(orders: List<OrderedItemModel>): LiveData<Resource<List<NewOrderModel>>> {
        return object : NetworkBoundResource<List<NewOrderModel>, List<NewOrderModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<NewOrderModel>>> {
                return RetrofitLiveData(mWebService.postActiveSessionOrders(orders))
            }
        }.asLiveData
    }

    fun postManageSessionOrders(sessionPk: Long, orders: List<OrderedItemModel>): LiveData<Resource<List<NewOrderModel>>> {
        return object : NetworkBoundResource<List<NewOrderModel>, List<NewOrderModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<NewOrderModel>>> {
                return RetrofitLiveData(mWebService.postSessionManagerOrders(sessionPk, orders))
            }
        }.asLiveData
    }

    fun postScheduledSessionOrders(sessionPk: Long, orders: List<OrderedItemModel>): LiveData<Resource<List<NewOrderModel>>> {
        return object : NetworkBoundResource<List<NewOrderModel>, List<NewOrderModel>>() {
            override fun shouldUseLocalDb(): Boolean {
                return false
            }

            override fun createCall(): LiveData<ApiResponse<List<NewOrderModel>>> {
                return RetrofitLiveData(mWebService.postScheduledSessionOrders(sessionPk, orders))
            }
        }.asLiveData
    }

    fun updateScheduledSessionOrder(sessionPk: Long, orderId: Long, quantity: Int): LiveData<Resource<OrderedItemModel>> {
        val body = Converters.objectMapper.createObjectNode().apply {
            put("quantity", quantity)
        }
        return object : NetworkBoundResource<OrderedItemModel, OrderedItemModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<OrderedItemModel>> = RetrofitLiveData(mWebService.patchScheduledSessionOrder(sessionPk, orderId, body))
        }.asLiveData
    }

    fun removeScheduledSessionOrder(sessionPk: Long, orderId: Long): LiveData<Resource<ObjectNode>> {
        return object : NetworkBoundResource<ObjectNode, ObjectNode>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<ObjectNode>> = RetrofitLiveData(mWebService.deleteScheduledSessionOrder(sessionPk, orderId))
        }.asLiveData
    }

    fun getRecommendedDishes(restaurantId: Long): LiveData<Resource<List<TrendingDishModel>>> {
        return object : NetworkBoundResource<List<TrendingDishModel>, List<TrendingDishModel>>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<List<TrendingDishModel>>> {
                return RetrofitLiveData(mWebService.getRestaurantRecommendedItems(restaurantId))
            }
        }.asLiveData
    }

    val checkCartStatus: LiveData<Resource<CartStatusModel>>
        get() = object : NetworkBoundResource<CartStatusModel, CartStatusModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun loadFromDb(): LiveData<CartStatusModel> = ObjectBoxInstanceLiveData(mCartBox.query().build())

            override fun createCall(): LiveData<ApiResponse<CartStatusModel>> = RetrofitLiveData(mWebService.sessionCartStatus)

            override fun saveCallResult(data: CartStatusModel?) {
                data?.let {
                    mCartBox.removeAll()
                    mCartBox.put(it)
                }
            }
        }.asLiveData

    val cartDetails: LiveData<Resource<CartDetailModel>>
        get() = object : NetworkBoundResource<CartDetailModel, CartDetailModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<CartDetailModel>> = RetrofitLiveData(mWebService.sessionCartDetail)
        }.asLiveData

    val cartBillData: LiveData<Resource<CartBillModel>>
        get() = object : NetworkBoundResource<CartBillModel, CartBillModel>() {
            override fun shouldUseLocalDb(): Boolean = false

            override fun createCall(): LiveData<ApiResponse<CartBillModel>> = RetrofitLiveData(mWebService.sessionCartBillDetail)
        }.asLiveData

    companion object : SingletonHolder<MenuRepository, Application>({ MenuRepository(it.applicationContext) })
}